source/lib.ts

```diff
@@ -0,0 +1,264 @@
+// /source/lib.ts
+// The option parser and rate limiting middleware
+
+import Express from 'express'
+
+import { Options, AugmentedRequest, RateLimitRequestHandler } from './types.js'
+import MemoryStore from './memory-store.js'
+
+/**
+ * Adds the defaults for options the user has not specified.
+ *
+ * @param options {Options} - The options the user specifies
+ *
+ * @returns {Options} - A complete configuration object
+ */
+const parseOptions = (passedOptions: Partial<Options>): Options => {
+	// Now add the defaults for the other options
+	const options: Options = {
+		windowMs: 60 * 1000,
+		store: new MemoryStore(passedOptions.windowMs ?? 60 * 1000),
+		max: 5,
+		message: 'Too many requests, please try again later.',
+		statusCode: 429,
+		headers: true,
+		useStandardizedHeaders: false,
+		requestPropertyName: 'rateLimit',
+		skipFailedRequests: false,
+		skipSuccessfulRequests: false,
+		requestWasSuccessful: (_request, response) => response.statusCode < 400,
+		keyGenerator: (request) => {
+			if (!request.ip) {
+				console.error(
+					'WARN | `express-rate-limit` | `request.ip` is undefined. You can avoid this by providing a custom `keyGenerator` function, but it may be indicative of a larger issue.',
+				)
+			}
+
+			return request.ip
+		},
+		skip: () => false,
+		handler: (_request, response, _optionsUsed: Options) =>
+			response.status(options.statusCode).send(options.message),
+		onLimitReached: () => {},
+		...passedOptions,
+	}
+
+	// Ensure that the store passed implements the {@link Store} interface
+	if (
+		typeof options.store.increment !== 'function' ||
+		typeof options.store.resetKey !== 'function' ||
+		(options.skipFailedRequests &&
+			typeof options.store.decrement !== 'function')
+	) {
+		throw new Error(
+			'An invalid store was passed. Please ensure that the store is a class that implements the `Store` interface.',
+		)
+	}
+
+	// Throw an error if any deprecated options are passed
+	const deprecatedOptions = ['global', 'delayMs', 'delayAfter'] as Array<
+		keyof Options
+	>
+	for (const option of deprecatedOptions) {
+		// This doesn't trigger if any value is set to a falsy value (e.g., 0),
+		// because that essentially disables them
+		if (passedOptions[option]) {
+			throw new Error(
+				`The \`${option}\` option is deprecated and will likely be removed from the \`express-rate-limit\` package in the future.`,
+			)
+		}
+	}
```

Comment:
I think we can get rid of these checks - but we're going to want to have similar code to check for the old headers names and recommend the new ones.